### PR TITLE
Add custom image URL

### PR DESCRIPTION
**Pull request by jamiehannaford**
Cherry picks 00a11ca from https://github.com/coreos/etcd-operator/pull/1056 and resolves merge conflicts.

Fixes https://github.com/coreos/etcd-operator/issues/1156.

### CODE REVIEW

File: pkg/spec/cluster.go [hunk 1]
```diff
@@ -70,6 +71,12 @@ type ClusterSpec struct {
 	// The vaild range of the size is from 1 to 7.
 	Size int `json:"size"`
 
+	// Image is the base etcd image name that will be used to launch
+	// etcd clusters. This is useful for private registries, etc.
+	//
+	// If image is not set, default is quay.io/coreos/etcd
+	Image string `json:"image"`
```
Comment (xiang90) [inline pkg/spec/cluster.go]: Image -> BaseImage?

File: pkg/spec/cluster.go [hunk 2]
```diff
@@ -26,6 +26,7 @@ import (
 )
 
 const (
+	defaultImage   = "quay.io/coreos/etcd"
```
Comment (xiang90) [inline pkg/spec/cluster.go]: defaultBaseImage

File: pkg/util/k8sutil/k8sutil.go [hunk 3]
```diff
@@ -109,8 +109,8 @@ func makeRestoreInitContainerSpec(backupAddr, token, version string, m *etcdutil
 	return string(b)
 }
 
-func EtcdImageName(version string) string {
-	return fmt.Sprintf("quay.io/coreos/etcd:v%v", version)
+func EtcdImageName(imageName, version string) string {
```
Comment (xiang90) [inline pkg/util/k8sutil/k8sutil.go]: baseImage, version string

**Comment by xiang90** [comment]
Can you also add this into CHANGELOG?

**Comment by xiang90** [comment]
LGTM.

defer to @hongchaodeng

File: CHANGELOG.md [hunk 4]
```diff
@@ -232,4 +236,3 @@ etcd container with this new field.
 ### Deprecated
 
 ### Security
-
```
Comment (hongchaodeng) [inline CHANGELOG.md]: Revert this change

File: CHANGELOG.md [hunk 5]
```diff
@@ -12,6 +12,10 @@
 
 ### Security
 
+## [Release 0.3.3]
```
Comment (hongchaodeng) [inline CHANGELOG.md]: Don't change this.

File: CHANGELOG.md [hunk 6]
```diff
@@ -12,6 +12,10 @@
...
+
+Adds ability for users to specify base image for etcd pods in a cluster.
```
Comment (hongchaodeng) [inline CHANGELOG.md]: Put this in a line under:
```
### Added

- Adds ...
```

File: pkg/spec/cluster.go [hunk 7]
```diff
@@ -70,6 +71,12 @@ type ClusterSpec struct {
 	// The vaild range of the size is from 1 to 7.
 	Size int `json:"size"`
 
+	// BaseImage is the base etcd image name that will be used to launch
+	// etcd clusters. This is useful for private registries, etc.
+	//
+	// If image is not set, default is quay.io/coreos/etcd
+	BaseImage string `json:"image"`
```
Comment (hongchaodeng) [inline pkg/spec/cluster.go]: json tag: baseImage

**Comment by xiang90** [comment]
@jamiehannaford kindly ping.

**Comment by jamiehannaford** [comment]
@xiang90 @hongchaodeng I've made some changes, please let me know if you need any more before release

File: CHANGELOG.md (hunk 4 above)
Comment (xiang90) [inline CHANGELOG.md]: @jamiehannaford can you revert this change?

**Comment by xiang90** [comment]
@jamiehannaford one minor issue. we can get this merged after fixing it.

File: pkg/util/k8sutil/k8sutil.go [hunk 8]
```diff
@@ -109,8 +109,8 @@ func makeRestoreInitContainerSpec(backupAddr, token, version string, m *etcdutil
 	return string(b)
 }
 
-func EtcdImageName(version string) string {
-	return fmt.Sprintf("quay.io/coreos/etcd:v%v", version)
+func EtcdImageName(baseImage, version string) string {
```
Comment (hongchaodeng) [inline pkg/util/k8sutil/k8sutil.go]: It doesn't make sense to call this `EtcdImageName` anymore.
Can you change it to `imageName`?

File: pkg/util/k8sutil/self_hosted.go [hunk 9]
```diff
@@ -69,7 +69,8 @@ func NewSelfHostedEtcdPod(m *etcdutil.Member, initialCluster, endpoints []string
 	}
 
 	commands = fmt.Sprintf("sleep 5; flock %s -c \"%s\"", etcdLockPath, commands)
-	c := etcdContainer(commands, cs.Version)
+	c := etcdContainer(commands, cs.BaseImage, cs.Version)
+
```
Comment (hongchaodeng) [inline pkg/util/k8sutil/self_hosted.go]: Can you revert the extra line space change?

File: pkg/util/k8sutil/pod_util.go [hunk 10]
```diff
@@ -34,11 +34,11 @@ func etcdVolumeMounts() []v1.VolumeMount {
 	}
 }
 
-func etcdContainer(commands, version string) v1.Container {
+func etcdContainer(commands, imageName, version string) v1.Container {
```
Comment (hongchaodeng) [inline pkg/util/k8sutil/pod_util.go]: imageName -> baseImage

File: pkg/spec/cluster.go [hunk 11]
```diff
@@ -188,6 +195,11 @@ func (c *ClusterSpec) Cleanup() {
 	if len(c.Version) == 0 {
 		c.Version = defaultVersion
 	}
+
+	if len(c.BaseImage) == 0 {
```
Comment (hongchaodeng) [inline pkg/spec/cluster.go]: Can you put this change before `if len(c.Version) == 0` or after `c.Version =` ?

**Comment by hongchaodeng** [comment]
Can you rebase master? We fixed `e2eslow` just a day ago.

**Comment by hongchaodeng** [comment]
LGTM

**Comment by jamiehannaford** [comment]
👍  Thanks!

**Comment by hongchaodeng** [comment]
Thanks for the work too
